src/dvc_data/hashfile/gc.py

```diff
@@ -10,6 +23,8 @@ def gc(odb, used, jobs=None, cache_odb=None, shallow=True):
         cache_odb = odb
     used_hashes = set()
     for hash_info in used:
+        if hash_info.name != odb.hash_name:
+            continue
```

Comment:
Previously we assumed we never had to check hash_name because we were only dealing with `md5`. In 3.x we need to check the hash name because otherwise hash values can overlap between `md5` and `md5-dos2unix`